app/scripts/modules/amazon/src/serverGroup/configure/wizard/pages/fields/ServerGroupDetailsField.tsx

```diff
@@ -0,0 +1,229 @@
+import { FormikProps } from 'formik';
+import { times } from 'lodash';
+
+import { HelpField, SETTINGS } from '@spinnaker/core';
+import { IAmazonServerGroupCommand } from '../../../serverGroupConfiguration.service';
+import { ParsedShards, parse, toFreeformDetails } from '../../../ServerGroupShardingUtils';
+
+const SUPPORTED_SHARD_DIMENSION_COUNT = SETTINGS.feature.shards ? 2 : 0;
+
+const removeSPELExpressions = (field: string) => field.split(/\$\{.*\}/).join('');
+
+const validateShard = (shard: string) =>
+  /^[a-zA-Z0-9]*$/.test(removeSPELExpressions(shard))
+    ? null
+    : 'Only alpha-numeric characters are allowed in the Shard field.';
+
+const validateOtherDetails = (detail: string) => {
+  const detailWithoutSPEL = removeSPELExpressions(detail);
+
+  if (!/^([a-zA-Z0-9._-])*$/.test(detailWithoutSPEL)) {
+    return 'Only dot(.), underscore(_), and dash(-) special characters are allowed in the Detail field.';
+  }
+
+  if (/-x\d/.test(`-${detailWithoutSPEL}`)) {
+    return 'Tokens of the form `-x\\d` cannot be used in the Detail field.';
+  }
+
+  return null;
+};
+
+interface LocalError {
+  error: string | null;
+  value: string | null;
+}
+/**
+ * This react hook is expected to listen to value changes and validate them. If there are no errors, then
+ * it will call the `successHandler`. If not, it will store error & value locally and expose it to the
+ * component without calling the `successHandler`. This is needed in components that merely render values that are
+ * completely controlled by their parent but wishes not to update the parent when there is an error.
+ *
+ * @param validator - A function that accepts the new value and returns a string if there is an error.
+ * @param successHandler - A function that is called by the hook when a new valid value is available.
+ * @param errorHandler - A function that is called by the hook when a new invalid value is available.
+ *
+ * Why keep track of a local error state?
+ *
+ * Values for `shards` and `other` are not stored as individual fields in the server group. Instead they are embedded
+ * into `freeformDetails` field along with other details. See `serverGroupShardingUtils` for more details on the
+ * format of the data.
+ *
+ * When the user is editing the value of this field, it is possible to introduce incorrect values such as hyphens,
+ * spaces, dots etc. These incorrect values will get combined to form the consolidated value for `freeformDetails`
+ * field. This consolidated value will be parsed again when `ServerGroupDetailsField` is re-rendered, but the parsing
+ * will fail this time since the consolidated value does not conform to the expected format. This will result in a
+ * state where the error cannot be shown to the user and won't be provided a way to fix it.
+ *
+ * Keeping a local error state allows us to bail out from informing the parent about the change and provide the user
+ * a way to fix the error.
+ */
+function useLocalError(
```

Comment:
This hook would be valuable as a reusable hook. If you move it to the `hooks` folder in `core/presentation` other fields with this same subtlety can utilize this hook.

Reply:
Thanks, I wasn't sure of the usefulness of this hook outside of this use case and was hesitant to move it out of the file. But if you feel this is useful, I will definitely move it out after making a few changes to make it a bit generic.